### PR TITLE
Add KH instability analytic data for GRMHD

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

Adds the KH instability initial data and executable.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by nilsdeppe** [comment]
Marking as priority since we need this for the referee report

File: src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.hpp [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+  /// The characteristic length for the width of the perturbation
+  struct PerturbWidth {
+    using type = double;
+    static type lower_bound() { return 0.0; }
+    static constexpr Options::String help = {
+        "The characteristic length for the width of the perturbation."};
+  };
+
+  /// The uniform magnetic field
+  struct MagneticField {
```
Comment (yoonso0-0) [inline src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.hpp:156]: Please mention magnetic field at the analytic data description in docs.

File: src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.hpp [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    static constexpr Options::String help = {"The uniform magnetic field."};
+  };
+
+  using options = tmpl::list<AdiabaticIndex, StripBimedianHeight,
+                             StripThickness, StripDensity, StripVelocity,
+                             BackgroundDensity, BackgroundVelocity, Pressure,
+                             PerturbAmplitude, PerturbWidth, MagneticField>;
+
+  static constexpr Options::String help = {
+      "Initial data to simulate the KH instability."};
```
Comment (yoonso0-0) [inline src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.hpp]: Add "magnetized"?

File: src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.cpp [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+template <typename DataType>
+tuples::TaggedTuple<hydro::Tags::RestMassDensity<DataType>>
+KhInstability::variables(
+    const tnsr::I<DataType, 3>& x,
+    tmpl::list<hydro::Tags::RestMassDensity<DataType>> /*meta*/) const {
+  auto result = make_with_value<Scalar<DataType>>(x, 0.0);
+  const size_t n_pts = get_size(get<0>(x));
+  for (size_t s = 0; s < n_pts; ++s) {
+    get_element(get(result), s) =
+        abs(get_element(get<2>(x), s) - strip_bimedian_height_) <
```
Comment (yoonso0-0) [inline src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.cpp:83]: Need to use `get<1>` to retrieve y coordinates

File: src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.cpp [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+tuples::TaggedTuple<hydro::Tags::SpatialVelocity<DataType, 3>>
+KhInstability::variables(
+    const tnsr::I<DataType, 3>& x,
+    tmpl::list<hydro::Tags::SpatialVelocity<DataType, 3>> /*meta*/) const {
+  auto result = make_with_value<tnsr::I<DataType, 3, Frame::Inertial>>(x, 0.0);
+
+  const size_t n_pts = get_size(get<0>(x));
+  for (size_t s = 0; s < n_pts; ++s) {
+    get_element(get<0>(result), s) =
+        abs(get_element(get<2>(x), s) - strip_bimedian_height_) <
```
Comment (yoonso0-0) [inline src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.cpp:120]: Same as above

File: src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.cpp [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+            ? strip_velocity_
+            : background_velocity_;
+  }
+
+  const double one_over_two_sigma_squared = 0.5 / square(perturbation_width_);
+  const double strip_lower_bound =
+      strip_bimedian_height_ - strip_half_thickness_;
+  const double strip_upper_bound =
+      strip_bimedian_height_ + strip_half_thickness_;
+  get<2>(result) =
```
Comment (yoonso0-0) [inline src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.cpp:131]: Same as above (all `get<2>`s from line 131 to 134)

File: src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.cpp [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+          const tnsr::I<DTYPE(data), 3, Frame::Inertial>& x, \
+          tmpl::list<TAG(data) < DTYPE(data)> >) const;
+
+GENERATE_INSTANTIATIONS(
+    INSTANTIATE_SCALARS, (double, DataVector),
+    (hydro::Tags::RestMassDensity, hydro::Tags::SpecificInternalEnergy,
+     hydro::Tags::Pressure, hydro::Tags::DivergenceCleaningField,
+     hydro::Tags::SpecificEnthalpy, hydro::Tags::LorentzFactor))
+
+#define INSTANTIATE_VELOCITY(_, data)                                         \
```
Comment (yoonso0-0) [inline src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.cpp]: Maybe change macro name to `INSTANTIATE_VECTORS`?

File: src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.cpp [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+  template tuples::TaggedTuple<TAG(data) < DTYPE(data), 3, Frame::Inertial> > \
+      KhInstability::variables(                                               \
+          const tnsr::I<DTYPE(data), 3, Frame::Inertial>& x,                  \
+          tmpl::list<TAG(data) < DTYPE(data), 3, Frame::Inertial> >) const;
+
+GENERATE_INSTANTIATIONS(INSTANTIATE_VELOCITY, (double, DataVector),
+                        (hydro::Tags::SpatialVelocity,
+                         hydro::Tags::MagneticField))
+}  // namespace grmhd::AnalyticData
```
Comment (yoonso0-0) [inline src/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.cpp:231]: Please add `#undef`s for macro definitions

File: tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+import numpy as np
+
+
+def rest_mass_density(x, adiabatic_index, strip_bimedian_height,
+                      strip_thickness, strip_density, strip_velocity,
+                      background_density, background_velocity, pressure,
+                      perturbation_amplitude, perturbation_width,
+                      magnetic_field):
+    return strip_density if np.absolute(
+        x[x.size - 1] -
```
Comment (yoonso0-0) [inline tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.py:13]: `x.size-1` -> `1`

File: tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+             background_velocity, pressure, perturbation_amplitude,
+             perturbation_width, magnetic_field):
+    dim = x.size
+    result = np.zeros(dim)
+    result[0] = strip_velocity if (
+        np.absolute(x[x.size - 1] - strip_bimedian_height) < 0.5 *
+        strip_thickness) else background_velocity
+    strip_lower_bound = strip_bimedian_height - 0.5 * strip_thickness
+    strip_upper_bound = strip_bimedian_height + 0.5 * strip_thickness
+    result[dim - 1] = (
```
Comment (yoonso0-0) [inline tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.py:28]: All `dim - 1` s  -> `1`

File: tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+
+def velocity(x, adiabatic_index, strip_bimedian_height, strip_thickness,
+             strip_density, strip_velocity, background_density,
+             background_velocity, pressure, perturbation_amplitude,
+             perturbation_width, magnetic_field):
+    dim = x.size
+    result = np.zeros(dim)
+    result[0] = strip_velocity if (
+        np.absolute(x[x.size - 1] - strip_bimedian_height) < 0.5 *
```
Comment (yoonso0-0) [inline tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.py:24]: `x.size - 1` -> `1`

File: tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        strip_thickness) else background_velocity
+    strip_lower_bound = strip_bimedian_height - 0.5 * strip_thickness
+    strip_upper_bound = strip_bimedian_height + 0.5 * strip_thickness
+    result[dim - 1] = (
+        np.exp(-0.5 *
+               ((x[dim - 1] - strip_lower_bound) / perturbation_width)**2) +
+        np.exp(-0.5 *
+               ((x[dim - 1] - strip_upper_bound) / perturbation_width)**2))
+    result[dim - 1] *= perturbation_amplitude * np.sin(4 * np.pi * x[0])
```
Comment (yoonso0-0) [inline tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.py:33]: Here as well

File: tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    return result
+
+
+def specific_internal_energy(x, adiabatic_index, strip_bimedian_height,
+                             strip_thickness, strip_density, strip_velocity,
+                             background_density, background_velocity, pressure,
+                             perturbation_amplitude, perturbation_width,
+                             magnetic_field):
+    return pressure / (adiabatic_index - 1.0) / strip_density if np.absolute(
+        x[x.size - 1] -
```
Comment (yoonso0-0) [inline tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/KhInstability.py:43]: `x.size - 1` -> `1`

File: tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/Test_KhInstability.cpp [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+}
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.PointwiseFunctions.AnalyticData.GrMhd.KhInstability",
+                  "[Unit][PointwiseFunctions]") {
+  pypp::SetupLocalPythonEnvironment local_python_env{
+      "PointwiseFunctions/AnalyticData/GrMhd"};
+
+  test(std::numeric_limits<double>::signaling_NaN());
+  // test(DataVector(5));
```
Comment (yoonso0-0) [inline tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/Test_KhInstability.cpp]: Q. Any reason why test for `DataVector` was skipped?

**Review by yoonso0-0 — CHANGES_REQUESTED** [review]
Please squash.

File: tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/Test_KhInstability.cpp (hunk 13 above)
Comment (nilsdeppe) [reply tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/Test_KhInstability.cpp]: Left over from debugging. Thanks for catching :)

**Comment by nilsdeppe** [comment]
@yoonso0-0 I've squashed in the fixes to your comments. I also changed the direction the strip is in (y instead of z) because this meant I could easily re-use ParaView scripts I have from other 2d simulations of GRMHD. I think that'll generally be useful

**Comment by yoonso0-0** [comment]
Great :)

**Review by kidder — APPROVED** [review]

**Pull request merged**